### PR TITLE
Adding in cache helpers

### PR DESCRIPTION
**Pull request by StabbyCutyou**
ping @jandre @jonschipp 

This adds the cache helpers.

Please pay special attention to the test binary I wrote to verify the file locker is as race-condition-free as I possibly could, and let me know if you see any issues.

### CODE REVIEW

**Comment by StabbyCutyou** [comment]
Ping @rossnanop

File: plugin/cache/cache.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+func (e InvalidCacheFileName) Error() string {
+	return string(e)
+}
+
+// OpenCacheFile will load the provided file from /var/cache/* and return a pointer to the
+// file if found, or an error if not found / something went wrong when opening. the name
+// argument should not begin with a slash, and should assume it will be appended to /var/cache
+// The caller is responsible for closing the file. If they don't, there could be problems.
+func OpenCacheFile(name string) (*os.File, error) {
+	if err := isReservedName(name); err != nil {
```
Comment (jandre) [inline plugin/cache/cache.go]: not for now, but both this and the python api should allow you to pass in file modes / perms optionally

**Comment by jandre** [comment]
looks fine

File: plugin/cache/cache.go (hunk 1 above)
Comment (StabbyCutyou) [inline plugin/cache/cache.go]: I'd say if you need special mode/perms, you can hit the fs directly. But that's an argument for another day :)

Comment (StabbyCutyou) [inline plugin/cache/cache.go]: Actually I could just public export the open call, it already does that

**Comment by StabbyCutyou** [comment]
What's the release process on the sdk?

**Comment by jandre** [comment]
Right now? merge and tag 

versioning is slightly weird, 0.2x will get built to the docker 0.2 tag that all plugins inherit from.  So, keep it a 0.2.x version even though this adds functionality.  we should probably bump it a major version and make the docker tag 1.x, but we have to change dockerhub to do all that along with the codegen, so we haven't yet.

**Comment by StabbyCutyou** [comment]
I'll increase the patch level, but I'd vote that classically this would be a minor bump, non breaking changes / new features

**Comment by StabbyCutyou** [comment]
(By which I mean, 0, 3 instead of 1.0)

**Comment by StabbyCutyou** [comment]
I'm gonna export the the low level open call with both mode/perms as arga before merging, it'll take 5 minutes and then we have more flexibility in the sdk :)

File: plugin/cache/cache.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+// was successful or not. In the event it was not, an error may or may not be returned (always check the value first
+// to know if it worked)
+// the name argument should not begin with a slash, and should assume it will be appended to /var/cache
+func LockCacheFile(name string) (bool, error) {
+	name = lockDir + stripLeftSlash(name)
+	var ok bool
+	var err error
+	for {
+		// Spin wait until something errors, or the file becomes free
+		if ok, err = utils.DoesFileExist(name); ok && err == nil {
```
Comment (rossnanop) [inline plugin/cache/cache.go]: I might golf this to check the error first:
```go
if ok, err = utils.DoesFileExist(name); err != nil {
  return false, err
}

// If lock file exists, wait until it becomes free
if ok {
  time.Sleep(lockWaitBackoff)
  continue
}
```

File: plugin/cache/cache.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+		if err != nil {
+			return false, err
+		}
+		// attempt an exclusive lock - if something already grabbed the file out from under us, we simply go back to waiting
+		var f *os.File
+		if f, err = openExclusiveFile(name); err != nil {
+			if os.IsExist(err) {
+				continue // The error was that the file existed - so we just keep on a'rollin
+			}
+			if err != nil {
```
Comment (rossnanop) [inline plugin/cache/cache.go]: I think you can remove this if statement.  If you are in this block, you know err != nil.  

```go
if os.IsExist(err) {
    continue // The error was that the file existed - so we just keep on a'rollin
}
return false, err
```

File: plugin/utils/file.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+// OpenFile is a low level call to safely open a file, with the provided flags and permissions
+// if the path underwhich the file is found does not exist, OpenFile will create that path
+// using os.ModePerm. However the behavior of whether or not it creates the file when not found
+// is up to the caller of OpenFile to determine via the flags/perms arguments. You can combine several
+// flags together, by passing them in a bitwise OR, ie: OpenFile("blah",os.O_RDWR|os.O_CREATE|os.O_EXCL,0600)
+func OpenFile(name string, flags int, perms os.FileMode) (*os.File, error) {
+	if _, err := os.Stat(name); err != nil {
+		// If it didn't exist
+		var ok bool
+		if ok, err = DoesFileExist(name); !ok && err == nil {
```
Comment (rossnanop) [inline plugin/utils/file.go]: Golf again, check error first:
```go
if ok, err = DoesFileExist(name); err != nil {
  return nil, err
}

// If file didn't exist, create the file and all directories leading up to it
if !ok {
    if err := os.MkdirAll(filepath.Dir(name), os.ModePerm); err != nil {
      return nil, err
    }
}
```

**Comment by StabbyCutyou** [comment]
Alright I exported the `open` call as`utils.OpenFile` and it takes flags/perms as an argument, but doesn't do any special checking - if you hand it some dumb shit, it will error at you. The cache calls mostly do the special checking, and are scoped exclusively to /var/cache.

So, for cache stuff, you always use `cache`

To do your own FS access in a mostly buyer-beware fashion, you can use `utils.OpenFile`, which is what `cache` relies upon.

File: plugin/cache/cache.go (hunk 3 above)
Comment (StabbyCutyou) [inline plugin/cache/cache.go]: Ah, nice catch - I think I left this around from an earlier attempt.

File: plugin/cache/cache.go (hunk 2 above)
Comment (StabbyCutyou) [inline plugin/cache/cache.go]: I suppose it is "more go" to check the error first, even though the behavior is ultimately the same. I'll revise it.

**Comment by StabbyCutyou** [comment]
alright so i was making sure my tests were getting run, and for some reason `make test` is skipping anything that is in a subdirectory of the `plugin` package. It's a simple fix, BUT: some of the tests that haven't been getting run, are failing :(

I'm gonna see about fixing them and making `make test` run all of them

**Comment by StabbyCutyou** [comment]
ping @jandre @rossnanop @jonschipp 

I made a few final changes, mostly relating to fixing the tests (not all running, some broken). Final LGTM sought.

Will go update the sphinx docs now for the new go methods.

**Comment by rossnanop** [comment]
LG(enough)TM. 😝 (but really my stuff was addressed, so I'm good with this)

**Comment by StabbyCutyou** [comment]
MARGIN

**Pull request merged**